spring-rabbit/src/test/java/org/springframework/amqp/rabbit/listener/BlockingQueueConsumerTests.java

```diff
@@ -137,6 +154,53 @@ public void testPrefetchIsSetOnFailedPassiveDeclaration() throws IOException {
 		verify(channel).basicQos(20);
 	}
 
+	@SuppressWarnings("unchecked")
+	@Test
+	public void testRecoverAfterDeletedQueueAndLostConnection() throws Exception {
+		ConnectionFactory connectionFactory = mock(ConnectionFactory.class);
+		Connection connection = mock(Connection.class);
+		Channel channel = mock(Channel.class);
+		given(connectionFactory.createConnection()).willReturn(connection);
+		given(connection.createChannel(false)).willReturn(channel);
+		given(connection.isOpen()).willReturn(true);
+		given(channel.isOpen()).willReturn(true);
+		final AtomicInteger n = new AtomicInteger();
+		ArgumentCaptor<Consumer> consumerCaptor = ArgumentCaptor.forClass(Consumer.class);
+		final CountDownLatch consumerLatch = new CountDownLatch(2);
+		willAnswer(invocation -> {
+			consumerLatch.countDown();
+			return "consumer" + n.incrementAndGet();
+		}).given(channel).basicConsume(anyString(),
+				anyBoolean(), anyString(), anyBoolean(), anyBoolean(), anyMap(), consumerCaptor.capture());
+		willThrow(new IOException("Intentional cancel fail")).given(channel).basicCancel("consumer2");
+
+		final BlockingQueueConsumer blockingQueueConsumer = new BlockingQueueConsumer(connectionFactory,
+				new DefaultMessagePropertiesConverter(), new ActiveObjectCounter<BlockingQueueConsumer>(),
+				AcknowledgeMode.AUTO, false, 1, "testQ1", "testQ2");
+		final CountDownLatch latch = new CountDownLatch(1);
+		Executors.newSingleThreadExecutor().execute(() -> {
+			blockingQueueConsumer.start();
+			while (true) {
+				try {
+					blockingQueueConsumer.nextMessage(1000);
+				}
+				catch (ConsumerCancelledException e) {
+					latch.countDown();
+					break;
+				}
+				catch (ShutdownSignalException e) {
+				}
+				catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+				}
+			}
+		});
+		assertTrue(consumerLatch.await(10, TimeUnit.SECONDS));
+		Consumer consumer = consumerCaptor.getValue();
+		consumer.handleCancel("consumer1");
```

Comment:
Not sure where is recover and queue deletion, but, eh, at least this test confirms that we do:
```
if (message == null && this.cancelled.get()) {
	throw new ConsumerCancelledException();
}
```
automatically on only one canceled `Consumer`.
Plus we are now independent of the `IOException` during cancellation process.

OK. Merging since the core code looks stable.